Code to skip large single terms and report them

angrycub opened the pull request:
Since a term > approx 30k can cause merge-index to be unable to compact
due to an overflow bug, we can skip the killer term and issue an error
to the logs.

Commit:
Code to skip large single terms and report them

Since a term > approx 30k can cause merge-index to be unable to compact
due to an overflow bug, we can skip the killer term and issue an error
to the logs.

Commit:
Improved check to look at size of entire posting

Was originally only checking the size of Term.  A sufficiently large
combination of bucket name, key name, etc could still cause an overflow

Commit:
Removed `GeneratedObject` from log message

It was just too noisy and gross.  Ended up emitting > 32000 bytes to log
when it was there which was truncated by lager.

jonmeredith commented on src/riak_indexed_doc.erl:
```diff
@@ -92,7 +92,15 @@ postings(IdxDoc) ->
                               Props = build_props(Positions, InlineFields),
-                              [{DocIndex, FieldName, Term, DocId, Props, K} | Acc]
+                              GeneratedTerm = {DocIndex, FieldName, Term, DocId, Props, K},
+                              TermLength = byte_size(term_to_binary(GeneratedTerm)),
```
more efficient to use `erlang:external_size(GeneratedTerm)`

Commit:
Changed sizing method

Originally used byte_length(to byte_size(term_to_binary(...)) which
causes the binary to be generated. Per Jon Meredith, switched to
erlang:external_size(...) which does the size estimation without
generating the binary

jonmeredith commented:
+1 merge
(I don't think Bors is operational here, or ever likely to be).

Tested by indexing this directory, containing a small, medium, big and huge.  The huge one had ~81k instances of the word 'huge' which is enough to overflow m_i.

```
jons-retina:riak(1.4) jmeredith$ ls -l /tmp/tosearch/
total 992
-rw-r--r--  1 jmeredith  wheel   57345 Apr 21 11:01 big
-rw-r--r--  1 jmeredith  wheel  409629 Apr 21 12:09 huge
-rw-r--r--  1 jmeredith  wheel   28521 Apr 21 11:02 medium
-rw-r--r--  1 jmeredith  wheel      30 Apr 21 11:00 small
jons-retina:riak(1.4) jmeredith$ wc -w /tmp/tosearch/*
   12288 /tmp/tosearch/big
   81926 /tmp/tosearch/huge
    5520 /tmp/tosearch/medium
       6 /tmp/tosearch/small
   99740 total
```

Seaching found all but huge

```
(riak@127.0.0.1)18> search:index_dir(<<"search">>, "/tmp/tosearch"). 
Starting...
Indexer <0.2269.0> - 100.0 % - 351.7 KB/sec - 4/4 files - 0 seconds
Finished.
ok
(riak@127.0.0.1)19> search:search(<<"search">>,<<"small">>).        
{1,
 [{<<"search">>,<<"small">>,
   [{p,[0]},{score,0.35355339059327373}]}]}
(riak@127.0.0.1)20> search:search(<<"search">>,<<"medium">>).       
{1,
 [{<<"search">>,<<"medium">>,
   [{p,[0,6,12,18,24,30,36,42,48,54,60,66,72,78,84,90,96,102,
        108,114|...]},
    {score,10.723805294763608}]}]}
(riak@127.0.0.1)22> search:search(<<"search">>,<<"huge">>).         
{0,[]}
(riak@127.0.0.1)23> search:search(<<"search">>,<<"big">>). 
{1,
 [{<<"search">>,<<"big">>,
   [{p,[0,6,12,18,24,30,36,42,48,54,60,66,72,78,84,90,96,102,
        108,114|...]},
    {score,16.0}]}]}
```

and the expected log entry was present in console.log

```
2014-04-21 14:07:45.853 [error] <0.28755.4>@riak_indexed_doc:postings:101 Encountered Large Posting (408942) when indexing Bucket "search", Key "huge", and Field Name "value", Skipping.
```

jonmeredith commented:
@jaredmorrow ping here so you can see riak_search fix for 2.0 or if we ever do a 1.4.9.

jaredmorrow commented:
Is there a PR to pull this into 2.0?

jonmeredith commented:
@jaredmorrow no specific PR, I've got branches with it merged to 1.4 and develop, will push them.
